src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/CertificateRequest.cs

```diff
@@ -530,20 +534,96 @@ public X509Certificate2 CreateSelfSigned(DateTimeOffset notBefore, DateTimeOffse
+                },
+                Validity = new ValidityAsn(notBefore, notAfter),
+                Subject = SubjectName.RawData,
             };
 
-            tbsCertificate.Extensions.AddRange(CertificateExtensions);
+            if (CertificateExtensions.Count > 0)
+            {
+                HashSet<string> usedOids = new HashSet<string>(CertificateExtensions.Count);
+                List<X509ExtensionAsn> extensionAsns = new List<X509ExtensionAsn>(CertificateExtensions.Count);
+
+                // An interesting quirk of skipping null values here is that
+                // Extensions.Count == 0 => no extensions
+                // Extensions.ContainsOnly(null) => empty extensions list
+
+                foreach (X509Extension extension in CertificateExtensions)
+                {
+                    if (extension == null)
+                    {
+                        continue;
+                    }
+
+                    if (!usedOids.Add(extension.Oid.Value))
+                    {
+                        throw new InvalidOperationException(
+                            SR.Format(SR.Cryptography_CertReq_DuplicateExtension, extension.Oid.Value));
+                    }
+
+                    extensionAsns.Add(new X509ExtensionAsn(extension, false));
+                }
+
+                tbsCertificate.Extensions = extensionAsns.ToArray();
+            }
+
+            using (AsnWriter writer = AsnSerializer.Serialize(tbsCertificate, AsnEncodingRules.DER))
+            {
+                byte[] encodedTbsCertificate = writer.Encode();                
+                CertificateAsn certificate = new CertificateAsn
+                {
+                    TbsCertificate = tbsCertificate,
+                    SignatureAlgorithm = signatureAlgorithmAsn,
+                    SignatureValue = generator.SignData(encodedTbsCertificate, HashAlgorithm)
+                };
 
-            return new X509Certificate2(tbsCertificate.Sign(generator, HashAlgorithm));
+                using (AsnWriter signedWriter = AsnSerializer.Serialize(certificate, AsnEncodingRules.DER))
+                {
+                    return new X509Certificate2(signedWriter.Encode());
+                }
+            }
+        }
+
+        private ReadOnlyMemory<byte> NormalizeSerialNumber(byte[] serialNumber)
+        {
+            if (serialNumber[0] >= 0x80)
+            {
+                // Keep the serial number unsigned by prepending a zero.
+                var newSerialNumber = new byte[serialNumber.Length + 1];
+                newSerialNumber[0] = 0;
+                serialNumber.CopyTo(newSerialNumber, 1);
+                return newSerialNumber;
+            }
+            else
+            {
+                // Strip any unnecessary zeros from the beginning.
+                int leadingZeros = 0;
+                while (leadingZeros < serialNumber.Length - 1 && serialNumber[leadingZeros] == 0 && serialNumber[leadingZeros + 1] < 0x80)
```

Comment:
I think the last clause will ArrayIndexOutOfBounds for all-zeros.

Reply:
The first clause should prevent that. Then again, it definitely feels hard to read.